### ISSUE TITLE
System logger in Windows logs everything with event id 0

### ISSUE THREAD

**Issue by tianweiliu** [cinder/Cinder#1109]
Background story is that I am trying to use a windows log agent(nxlog, rsyslog, etc.) to forward system logs created by my cinder app.
There are 2 parts of the problem I am facing:
* Without proper instrumentation manifest, Windows Event Viewer will generate following error message in event log:
```
The description for Event ID 0 from source XXX cannot be found. Either the component that raises this event is not installed on your local computer or the installation is corrupted. You can install or repair the component on the local computer.
```
I can see the raw data but when I use log agents, they only forward the error message.
I can fix this error by adding a new generic event source in the registry, but event viewers only reports the first line of the event data as message.(aka. the metadata portion) which is not really useful to me.

* The cinder event log's ```write``` function is hard coded to report all logs with event id 0:

As wrote in Log.cpp:
```
::ReportEventW( mHLog, eventLevel, 0, 0, 0, 2, 0, wStrings, 0 );
```
This is blocking me from writing proper instrumentation manifest xml for windows event log to recognize the format of cinder logs, as windows event log requires unique non-zero positive integer event ids for each custom log event (aka. different event levels).
As seen here:
http://blog.dlgordon.com/2012/06/writing-to-event-log-in-net-right-way.html

Worst case scenario I will have the log agent to read the file log.

**Comment by lucasvickers** [comment]
Thanks for referencing that article.  I'm going to give it a read then give you some feedback.

**Comment by lucasvickers** [comment]
Let's solve this for your manifest, and in doing so we can discuss how to create a generalized manifest file we can ship with Cinder.  

While on MSW we've been able to get events into the Event Log with minimum information, but WinRT has been more problematic and straight up rejecting the logging calls.  A proper manifest should fix this, but to be honest I wasn't having luck finding an easily digestible article on writing a program manifest, so again, thanks for the article!  In case your curious, #996 is the WinRT issue that I'm referencing.

> I can fix this error by adding a new generic event source in the registry, but event viewers only 
> reports the first line of the event data as message.(aka. the metadata portion)

In my call, ```wStrings``` holds two strings.  The first is the meta information, the second the error message.  I think once we get a proper manifest in place it should be easy to get both strings to show up.  You probably need to structure your event template (inside the manifest) to accept two strings.

#### Writing a / your Manifest
* **GUID** : Any sense on how we're supposed to obtain this?
* **Channel** : Seems straight forward
* **Levels** : We can probably use the standard windows ones, but the conversion is not one to one, as seen [here](https://github.com/cinder/Cinder/blob/master/src/cinder/Log.cpp#L531-L547).  I'm not concerned about the loss of granularity in INFO, DEBUG, and VERBOSE, because I mainly see system logs being used for error reporting (as you are doing), not viewing debug trace.
* **Tasks** : Default seems fine
* **Opcodes** : Default seems fine

#### Making the proper logging call
For MSW, we're currently using [ReportEventW](https://msdn.microsoft.com/en-us/library/windows/desktop/aa363679(v=vs.85).aspx).  This differs slightly form ```WriteEvent```.  I'm not entirely sure which one is 'correct', but I can't find any good documentation for ```WriteEvent```.

```ReportEventW``` does not have a channel param, and the API docs don't have any mention of a manifest.

I can easily change ```ReportEventW``` so that we have EventIDs that are associated with levels.  I could take the [enums](https://github.com/cinder/Cinder/blob/master/include/cinder/Log.h#L40) for log levels, and add 10 or 100 so they are non-zero, _but is this enough for you to write a proper manifest?_

#### Generalizing the manifest

I don't have much to say on this yet, but I want to learn from this, and see if we can't derive a default manifest file.  My biggest concern is that the GUID would always be the same, unless we got the TinderBox to generate a new GUID for us.  I don't yet understand where we're even *supposed* to get the GUID from -- any ideas?

#### Next steps

Give me your input and I can write up a test branch for us to work out of.  If we just want to try changing the EventIDs to correlate with log levels that's pretty easy.

**Comment by lucasvickers** [comment]
Please work out of this branch and let me know if it's enough to get a manifest working.

https://github.com/lucasvickers/Cinder/tree/windows-logging

**Comment by tianweiliu** [comment]
Thanks for the timely replies! 

Windows SDK ships with a manifest generator (ecmangen.exe) and we can generate a manifest file there. Here is a good article I found: http://blogs.msdn.com/b/seealso/archive/2011/06/08/use-this-not-this-logging.aspx

I will try to generate a manifest file this week and see if it works!

And once that is working I am pretty sure that file works with all cinder apps, and you guys can ship that as a generic manifest. The only thing that developer needs to do is to generate another GUID and change the source name to their app's name.

**Comment by tianweiliu** [comment]
Here is a workaround for nxlog reading file log instead of event log: https://gist.github.com/tianweiliu/846a48147f0b2acd5ffe

But I will still try to get the manifest working.

**Comment by lucasvickers** [comment]
Thanks.  
Yes please do try the manifest.

**Comment by tianweiliu** [comment]
Things are looking very promising now. 
The article that helped me the most: http://tomestephens.com/2013/03/writing-an-etw-event-provider-in-c/

Here are the steps to get it working:
# Create manifest
* Make sure you have Windows SDK installed, this can be done by modifying Visual Studio's installation in Program and Features.
* Go to `C:\Program Files (x86)\Windows Kits\(SDK version)\bin\(platform)\` and open *Manifest Generator* `ecmangen.exe`.
* I have created a simple generic manifest for Cinder as a starting point.
  * Download the manifest here: https://gist.github.com/tianweiliu/55c8a1246daa3de988a5
* Load the manifest in Manifest Generator and edit the event provider.
  * Set the `Name` to be the same as [stem (aka. event source name)](https://github.com/cinder/Cinder/blob/master/src/cinder/Log.cpp#L488).
  * Generate a new `GUID`.
  * Put the absolute path to executable in both `Resources` and `Messages` field. *Yes. You have to know where the final executable will be placed before you even deploy the application.*
  * As written [here] (https://github.com/cinder/Cinder/blob/master/src/cinder/Log.cpp#L520), there are two fields of data that cinder outputs, one is `mMeta` and the other is `mText`. I set up that as a template already in the manifest. (although i change the name from `mMeta` to `Metadata` and `mText` to `message`)
  * Change events to display message differently if you want, in my manifest I set it to be the same format as file logging. `%1` is metadata, `%2` is text. (for example: `%1 %2` will display `|warning| 2015-09-21.11:03:21 dotenv::defaultEnv[131] using default DEBUG settings.`)
  * Save the manifest file.

# Compiling the manifest
* Open *Command Prompt* and `cd` into `C:\Program Files (x86)\Windows Kits\(SDK version)\bin\(platform)\`
* Compile manifest you just created by running `mc.exe myCinderApp.man`
* You will get `myCinderApp.rc`, `MSG00001.bin`, `myCinderApp.h`, `myCinderAppTEMP.bin` in the bin folder.
* Copy/Move them to your project folder and import them into your project structure.
* Rebuild your cinder application.

# Register manifest
* Give USERS group permission to Read & Execute the complied exe.
* Open `Command Prompt` and run `wevtutil im myCinderApp.man`
* Run `wevtutil gp yourEventSourceName`. If you see cmd outputting a tree structure representing the custom event log you just created, you are good to go.
* Reopen Event Viewer and you should see event being parsed correctly.

**Comment by lucasvickers** [comment]
Glad to hear you got it working, and thanks for spelling all of these steps out for us!

I'll begin working on this in the near future, and it'll definitely find its way into the logging guide I'm working on.  In the mean time I'll put in a PR for the unique event IDs.

**Comment by richardeakin** [comment]
I believe this was fixed in #1125, let us know otherwise.

**Comment by lucasvickers** [comment]
Yeah it was fixed, but I want to use the information from this issue to feed into our WinRT logging work.  I guess there's no strong reason to keep the issue open.

@tianweiliu have things been working out for you?
